pkg/aws/ec2/ec2.go

```diff
@@ -223,87 +224,147 @@ func (c *Collector) fetchVolumesData(ctx context.Context, regionClient client.Cl
-					region = region[:len(region)-1]
-				}
+	// Process instances in parallel with limited concurrency
+	instanceChan := make(chan ec2Types.Instance, len(allInstances))
+	for _, instance := range allInstances {
+		instanceChan <- instance
+	}
+	close(instanceChan)
 
-				price, err := c.computePricingMap.GetPriceForInstanceType(region, string(instance.InstanceType))
-				if err != nil {
-					c.logger.Error(fmt.Sprintf("error getting price for instance type %s: %s", instance.InstanceType, err))
-					continue
-				}
+	// Limit concurrent instance processing
+	workers := maxConcurrentWorkers
+	if len(allInstances) < workers {
+		workers = len(allInstances)
+	}
 
-				labelValues := []string{
-					*instance.PrivateDnsName,
-					*instance.InstanceId,
-					region,
-					c.computePricingMap.InstanceDetails[string(instance.InstanceType)].InstanceFamily,
-					string(instance.InstanceType),
-					clusterName,
-					pricetier,
-					string(instance.Architecture),
-				}
-				ch <- prometheus.MustNewConstMetric(InstanceCPUHourlyCostDesc, prometheus.GaugeValue, price.Cpu, labelValues...)
-				ch <- prometheus.MustNewConstMetric(InstanceMemoryHourlyCostDesc, prometheus.GaugeValue, price.Ram, labelValues...)
-				ch <- prometheus.MustNewConstMetric(InstanceTotalHourlyCostDesc, prometheus.GaugeValue, price.Total, labelValues...)
+	instanceWaitGroup := sync.WaitGroup{}
+	for i := 0; i < workers; i++ {
+		instanceWaitGroup.Add(1)
+		go func() {
+			defer instanceWaitGroup.Done()
+			for instance := range instanceChan {
+				c.processInstance(instance, ch)
 			}
-		}
+		}()
+	}
+	instanceWaitGroup.Wait()
+}
+
+// processInstance processes a single instance and emits its metrics
+func (c *Collector) processInstance(instance ec2Types.Instance, ch chan<- prometheus.Metric) {
+	clusterName := client.ClusterNameFromInstance(instance)
+	if instance.PrivateDnsName == nil || *instance.PrivateDnsName == "" {
+		c.logger.Debug(fmt.Sprintf("no private dns name found for instance %s", *instance.InstanceId))
+		return
+	}
+	if instance.Placement == nil || instance.Placement.AvailabilityZone == nil {
+		c.logger.Debug(fmt.Sprintf("no availability zone found for instance %s", *instance.InstanceId))
+		return
+	}
+
+	region := *instance.Placement.AvailabilityZone
+
+	pricetier := "spot"
+	if instance.InstanceLifecycle != ec2Types.InstanceLifecycleTypeSpot {
+		pricetier = "ondemand"
+		// Ondemand instances are keyed based upon their Region, so we need to remove the availability zone
+		region = region[:len(region)-1]
+	}
+
+	price, err := c.computePricingMap.GetPriceForInstanceType(region, string(instance.InstanceType))
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("error getting price for instance type %s: %s", instance.InstanceType, err))
+		return
 	}
+
+	labelValues := []string{
+		*instance.PrivateDnsName,
+		*instance.InstanceId,
+		region,
+		c.computePricingMap.InstanceDetails[string(instance.InstanceType)].InstanceFamily,
```

Comment:
we need an RLock if we're going to read from the computePricingMap. We can create a getter function similar to to GetPriceForInstanceType() to get the instance family safely